lib/uplink/main.go

```diff
@@ -0,0 +1,32 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"storj.io/storj/pkg/storj"
+)
+
+// Caveats could be things like a read-only restriction, a time-bound
+// restriction, a bucket-specific restriction, a path-prefix restriction, a
+// full path restriction, etc.
+type Caveat interface {
+}
+
+// A Macaroon represents an access credential to certain resources
+type Macaroon interface {
+	Serialize() ([]byte, error)
+	Restrict(caveats ...Caveat) Macaroon
+}
+
+// Permissions give someone specific access.
+type Permissions struct {
+	Macaroon Macaroon
+}
+
+// Encryption specifies an individual bucket's encryption choices
+type Encryption struct {
+	Key           storj.Key
+	EncPathPrefix storj.Path
```

Comment:
What is this actually?

Reply:
if i give you access to just /documents/music, then you'll get the key for everything under /documents/music, and the encrypted path for /documents/music so you know what to prepend